pkgs/build-support/rust/fetch-cargo-vendor-util.py

```diff
@@ -0,0 +1,247 @@
+import functools
+import glob
+import hashlib
+import json
+import multiprocessing as mp
+import os
+import re
+import shutil
+import subprocess
+import sys
+
+import requests
+import tomli
+
+
+eprint = functools.partial(print, file=sys.stderr)
+
+
+def load_toml(path: str) -> dict:
+    with open(path, "rb") as f:
+        return tomli.load(f)
+
+
+def download_file_with_checksum(url: str, destination_path: str, expected_checksum: str):
+    """Downloads a file and computes its SHA-256 checksum while writing it to disk."""
+    sha256_hash = hashlib.sha256()
+    with requests.get(url, stream=True) as response:
+        if not response.ok:
+            raise Exception(f"Failed to download file from {url}. Status code: {response.status_code}")
+        with open(destination_path, "wb") as file:
+            for chunk in response.iter_content(1024):  # Download in chunks
+                if chunk:  # Filter out keep-alive chunks
+                    file.write(chunk)
+                    sha256_hash.update(chunk)
+
+    # Compute the final checksum
+    calculated_checksum = sha256_hash.hexdigest()
+
+    if calculated_checksum != expected_checksum:
+        raise Exception(f"Hash mismatch! File fetched from {url} had checksum {calculated_checksum}, expected {expected_checksum}.")
+
+
+def get_download_url_for_tarball(pkg: dict):
```

Comment:
```suggestion
def get_download_url_for_tarball(pkg: dict[str, Any]) -> str:
```

Reply:
I proposed to use `typeDict` above, which would provide way better typing